plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -19,74 +19,90 @@
+                    + Strings.collectionToDelimitedString(clientsSettings.keySet(), ","));
         }
-
-        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
-        HttpRequestInitializer requestInitializer =  createRequestInitializer(clientSettings);
-
-        Storage.Builder storage = new Storage.Builder(transport, JacksonFactory.getDefaultInstance(), requestInitializer);
-        if (Strings.hasLength(clientSettings.getApplicationName())) {
-            storage.setApplicationName(clientSettings.getApplicationName());
+        final NetHttpTransport netHttpTransport = GoogleNetHttpTransport.newTrustedTransport();
+        final HttpTransportOptions httpTransportOptions = HttpTransportOptions.newBuilder()
+                .setConnectTimeout(toTimeout(clientSettings.getConnectTimeout()))
+                .setReadTimeout(toTimeout(clientSettings.getReadTimeout()))
+                .setHttpTransportFactory(() -> netHttpTransport)
+                .build();
+        final StorageOptions.Builder storageOptionsBuilder = StorageOptions.newBuilder()
+                .setTransportOptions(httpTransportOptions)
+                .setHeaderProvider(() -> {
+                    final MapBuilder<String, String> mapBuilder = MapBuilder.newMapBuilder();
+                    if (Strings.hasLength(clientSettings.getApplicationName())) {
+                        mapBuilder.put("user-agent", clientSettings.getApplicationName());
+                    }
+                    return mapBuilder.immutableMap();
+                });
+        if (Strings.hasLength(clientSettings.getHost())) {
+            storageOptionsBuilder.setHost(clientSettings.getHost());
         }
-        if (Strings.hasLength(clientSettings.getEndpoint())) {
-            storage.setRootUrl(clientSettings.getEndpoint());
+        if (Strings.hasLength(clientSettings.getProjectId())) {
+            storageOptionsBuilder.setProjectId(clientSettings.getProjectId());
         }
-        return storage.build();
-    }
-
-    static HttpRequestInitializer createRequestInitializer(final GoogleCloudStorageClientSettings settings) throws IOException {
-        GoogleCredential credential = settings.getCredential();
-        if (credential == null) {
-            credential = GoogleCredential.getApplicationDefault();
+        if (clientSettings.getCredential() == null) {
+            logger.warn("\"Application Default Credentials\" are not supported out of the box."
+                    + " Additional file system permissions have to be granted to the plugin.");
+        } else {
+            final ServiceAccountCredentials serviceAccountCredentials = clientSettings.getCredential();
+            if (Strings.hasLength(clientSettings.getTokenUri().toString())) {
+                storageOptionsBuilder
+                        .setCredentials(serviceAccountCredentials.toBuilder().setTokenServerUri(clientSettings.getTokenUri()).build());
```

Comment:
override token uri, using client setting from `elasticsearch.yml`

Reply:
Maybe split this in multiple lines for a better readability?
```
            final URI tokenUri = clientSettings.getTokenUri();
            if (tokenUri != null) {
                // Rebuild the service account credentials in order to use a custom Token url.
                // This is mostly used for testing purpose.
                ServiceAccountCredentials.Builder credentialsBuilder = serviceAccountCredentials.toBuilder();
                credentialsBuilder.setTokenServerUri(tokenUri);
                storageOptionsBuilder.setCredentials(credentialsBuilder.build());
            } else {
                storageOptionsBuilder.setCredentials(serviceAccountCredentials);
            }
```

Reply:
hope this fills the bill:
```
            ServiceAccountCredentials serviceAccountCredentials = clientSettings.getCredential();
            // override token server URI
            final URI tokenServerUri = clientSettings.getTokenUri();
            if (Strings.hasLength(tokenServerUri.toString())) {
                // Rebuild the service account credentials in order to use a custom Token url.
                // This is mostly used for testing purpose.
                serviceAccountCredentials = serviceAccountCredentials.toBuilder().setTokenServerUri(tokenServerUri).build();
            }
            storageOptionsBuilder.setCredentials(serviceAccountCredentials);
```

Reply:
Thanks!